octopusdeploy_framework/schemas/step_template.go

```diff
@@ -0,0 +1,288 @@
+package schemas
+
+import (
+	"fmt"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	ds "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	rs "github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"regexp"
+)
+
+const (
+	StepTemplateResourceDescription   = "step_template"
+	StepTemplateDatasourceDescription = "step_template"
+)
+
+type StepTemplateTypeDataSourceModel struct {
+	ID           types.String `tfsdk:"id"`
+	SpaceID      types.String `tfsdk:"space_id"`
+	StepTemplate types.Object `tfsdk:"step_template"`
+}
+
+type StepTemplateTypeResourceModel struct {
+	ActionType                types.String `tfsdk:"action_type"`
+	SpaceID                   types.String `tfsdk:"space_id"`
+	CommunityActionTemplateId types.String `tfsdk:"community_action_template_id"`
+	Name                      types.String `tfsdk:"name"`
+	Description               types.String `tfsdk:"description"`
+	Packages                  types.List   `tfsdk:"packages"`
+	Parameters                types.List   `tfsdk:"parameters"`
+	Properties                types.Map    `tfsdk:"properties"`
+	StepPackageId             types.String `tfsdk:"step_package_id"`
+	Version                   types.Int32  `tfsdk:"version"`
+
+	ResourceModel
+}
+
+type StepTemplatePackageType struct {
+	ID                  types.String `tfsdk:"id"`
+	AcquisitionLocation types.String `tfsdk:"acquisition_location"`
+	Name                types.String `tfsdk:"name"`
+	FeedID              types.String `tfsdk:"feed_id"`
+	PackageID           types.String `tfsdk:"package_id"`
+	Properties          types.Object `tfsdk:"properties"`
```

Comment:
Just curious - any particular reason this properties is an `Object` vs. the step template properties which is a `Map`?

Reply:
Yes, we know the properties for a package, they are always the same so we can code the as a object so terraform tools will give you auto complete and the properties will show up in the docs.

We do not know the properties on the step template because it changes based on the step type so we use a generic bag (map) so it can work with any step type.